File: exercise/day24/src/main/java/M/Submarine.java
package M;

import java.util.List;

public class Submarine {
    private final Position position;

    private Submarine(Position position) {
        this.position = position;
    }

    public Submarine(int horizontal, int depth) {
        this(new Position(horizontal, depth));
    }

    public Submarine move(List<Instruction> instructions) {
        return instructions
                .stream()
                .reduce(this, Submarine::move, (p, n) -> p);
    }

    private Submarine move(Instruction instruction) {
        return new Submarine(switch (instruction.text()) {
            case "down" -> position.changeDepth(position.depth() + instruction.x());
            case "up" -> position.changeDepth(position.depth() - instruction.x());
            default -> position.moveHorizontally(position.horizontal() + instruction.x());
        });
    }

    public Position getPosition() {
        return position;
    }
}

File: exercise/c#/Day24/Day24/Submarine.cs
namespace Day24
{
    public class Submarine(Position position)
    {
        public Submarine(int horizontal = 0, int depth = 0)
            : this(new Position(horizontal, depth))
        {
        }

        public Submarine Move(IEnumerable<Instruction> instructions)
            => instructions
                .Aggregate(this, (submarine, instruction) => submarine.Move(instruction));

        private Submarine Move(Instruction instruction)
            => new(instruction.Text switch
            {
                "down" => position.ChangeDepth(position.Depth + instruction.X),
                "up" => position.ChangeDepth(position.Depth - instruction.X),
                _ => position.MoveHorizontally(position.Horizontal + instruction.X)
            });

        public Position Position() => position;
    }
}